pkg/controller/aws/network/securitygroup/controller.go

```diff
@@ -0,0 +1,227 @@
+	}
+
+	return &external{c}, nil
+}
+
+type external struct {
+	client ec2.SecurityGroupClient
+}
+
+func (e *external) Observe(ctx context.Context, mgd resource.Managed) (resource.ExternalObservation, error) {
+	cr, ok := mgd.(*v1alpha1.SecurityGroup)
+	if !ok {
+		return resource.ExternalObservation{}, errors.New(errUnexpectedObject)
+	}
+
+	// To find out whether a SecurityGroup exist:
+	// - the object's ExternalState should have securityGroupId populated
+	// - a SecurityGroup with the given securityGroupId should exist
+	if cr.Status.SecurityGroupID == "" {
+		return resource.ExternalObservation{
+			ResourceExists: false,
+		}, nil
+	}
+
+	req := e.client.DescribeSecurityGroupsRequest(&awsec2.DescribeSecurityGroupsInput{
+		GroupIds: []string{cr.Status.SecurityGroupID},
+	})
+	req.SetContext(ctx)
+
+	response, err := req.Send()
+
+	if err != nil {
+		if ec2.IsSecurityGroupNotFoundErr(err) {
+			return resource.ExternalObservation{
+				ResourceExists: false,
+			}, nil
+		}
+
+		return resource.ExternalObservation{}, errors.Wrapf(err, "failed to describe SecurityGroup with id: %v", cr.Status.SecurityGroupID)
+	}
+
+	// in a successful response, there should be one and only one object
+	if len(response.SecurityGroups) != 1 {
+		return resource.ExternalObservation{}, errors.Errorf("retrieved multiple SecurityGroups for the given securityGroupId: %v", cr.Status.SecurityGroupID)
+	}
+
+	observed := response.SecurityGroups[0]
+
+	// the fact that the security group is successfully fetched,
+	// indicates that it's available
+	cr.SetConditions(runtimev1alpha1.Available())
+
+	cr.Status.SecurityGroupExternalStatus = cr.BuildExternalStatusFromObservation(observed)
+
+	return resource.ExternalObservation{
+		ResourceExists:    true,
+		ConnectionDetails: resource.ConnectionDetails{},
+	}, nil
+}
+
+func (e *external) Create(ctx context.Context, mgd resource.Managed) (resource.ExternalCreation, error) {
+	cr, ok := mgd.(*v1alpha1.SecurityGroup)
+	if !ok {
+		return resource.ExternalCreation{}, errors.New(errUnexpectedObject)
+	}
+
+	cr.Status.SetConditions(runtimev1alpha1.Creating())
+
+	// Creating the SecurityGroup itself
+	req := e.client.CreateSecurityGroupRequest(&awsec2.CreateSecurityGroupInput{
+		GroupName:   aws.String(cr.Spec.GroupName),
+		VpcId:       aws.String(cr.Spec.VPCID),
+		Description: aws.String(cr.Spec.Description),
+	})
+	req.SetContext(ctx)
+
+	result, err := req.Send()
+	if err != nil {
+		return resource.ExternalCreation{}, errors.Wrap(err, "failed to create the SecurityGroup resource")
+	}
+
+	cr.Status.SecurityGroupExternalStatus = cr.BuildExternalStatusFromObservation(awsec2.SecurityGroup{GroupId: result.GroupId})
+
+	if err != nil {
+		return resource.ExternalCreation{}, errors.Wrap(err, "failed to create the SecurityGroup resource")
+	}
+
+	// Authorizing Ingress permissions for the SecurityGroup
+	ingressPerms := cr.BuildEC2Permissions(cr.Spec.IngressPermissions)
```

Comment:
Would it be a higher fidelity representation of the AWS API to model these ingress and egress permissions as their own managed resources, rather than as fields of the security group managed resource?

Reply:
I don't think it limits the fidelity. As it's been the case in tools like Terraform https://www.terraform.io/docs/providers/aws/r/security_group.html, I figured grouping all these resources into one resource is more straightforward as one could define all these related resources in one object